Added USBGecko Support, other fixes

techydude0713 opened the pull request:
**This PR adds USBGecko connectability to OSCDL.** 
A new interface was created for the "Send to Wii" dialog that features a IP address text box, and a serial device combobox.
<img width="350" alt="TCPIPMenu" src="https://user-images.githubusercontent.com/39745188/200706518-843d9d03-56fa-4f63-ac9e-6f4017de2ef9.png"><img width="350" alt="USBGeckoMenu" src="https://user-images.githubusercontent.com/39745188/200706565-360ed441-ed02-4043-8e1d-059b7d8b7b9a.png">

With the USBGecko, if a device is plugged in and detected, OSCDL will automatically select the device in the combobox.

Other fixes:
+ The chunk size was lowered to have a stable transmission. (When too high, the content fails to send.)
+ Fixed icons_not loading due to Threading issue (load_icon takes one argument, but was supplied two).
+ ~In the CLI: "send" is now split to "ipsend" and "geckosend."~ Now simplified to send.
+ Enabled safe mode for Send to Wii
+ Prioritize Send to Wii status messages over icon loading (caused progress bar to be -1)
+ Fixed unsendable Wii applications having the "Send to Wii" toggle on after pressing the Download button.
+ CLI: If no arguments, display the help message.

Note: this adds the requirement of PySerial.

dhtdht020 commented:
Looks very promising, thank you so much.
Will look into this PR in a few days, I'm a little busy right now.

ed0nkey commented:
Just tried this, it sends the data but then HBC errors out with "error while decompressing received data"
![photo_2022-11-09_20-26-13](https://user-images.githubusercontent.com/79872165/200923258-a6a87e47-e99e-41d7-9186-add674b9bbe4.jpg)

techydude0713 commented:
@ed0nkey 
It should be fixed. Could you try again with the latest commit?

ed0nkey commented:
> @ed0nkey It should be fixed. Could you try again with the latest commit?

it works now 👍

dhtdht020 commented on gui/ui_wiiloaddialog.ui:
```diff
@@ -0,0 +1,195 @@
+        <attribute name="title">
+         <string>TCP/IP</string>
+        </attribute>
+        <widget class="QPlainTextEdit" name="IPBox">
+         <property name="geometry">
+          <rect>
+           <x>270</x>
+           <y>280</y>
+           <width>181</width>
+           <height>31</height>
+          </rect>
+         </property>
+        </widget>
```
The widget that should be used for simple single line input is QLineEdit rather than QPlainTextEdit, which is intended for advanced text editing. Switching to QLineEdit might spare you from some of the handling and event cancellations for the IPBox you've done in code.

dhtdht020 commented on README.md:
```diff
@@ -5,6 +5,8 @@ OSCDL is a cross platform desktop client for the Open Shop Channel homebrew repo
 
 With OSCDL, you can download hundreds of homebrew apps and themes to your computer or wirelessly send them directly to the Wii.
 
+OSCDL also supports USBGecko connections if the Wii cannot connect via LAN.
```
```suggestion
With OSCDL, you can download hundreds of homebrew apps and themes to your computer send them directly to the Wii through the network or through USB Gecko.
```

dhtdht020 commented on osc-dl.py:
```diff
@@ -30,8 +33,10 @@
 get = subparser.add_parser(name='get', help="Download application.")
-# Send - downloads and sends application
-send = subparser.add_parser(name='send', help="Send application to Wii.")
+# IPSend - downloads and sends application over TCP/IP
+ipsend = subparser.add_parser(name='ipsend', help="Send application to Wii via TCP/IP.")
+# GeckoSend - downloads and sends application over USBGecko
+geckosend = subparser.add_parser(name='geckosend', help="Send application to Wii via USBGecko protocol.")
```
It's probably better to keep the command "send" as is, for backwards compatibility purpose.
You could maybe make geckosend a mode of send for when something along the lines of "--gecko" is specified, or just rename ipsend back to just send.

dhtdht020 commented on gui_helpers.py:
```diff
@@ -1,3 +1,29 @@
+from PySide6.QtWidgets import QMessageBox
 
 settings = QSettings("Open Shop Channel", "OSCDL")
+
+DATASENT = True
+
+def sendGecko(data,conn,window,path_to_app):
+    try:
+        conn.send(data)
+    except Exception as e:
+        logging.error('Error while connecting to the HBC. Close any dialogs on HBC and try again.')
+        QMessageBox.warning(window, 'Connection error',
+            'Error while connecting to the HBC. Close any dialogs on HBC and try again.')
+        print(f'WiiLoad: {e}')
+        window.ui.progressBar.setValue(0)
+        window.status_message('Error: Could not connect to the Homebrew Channel. :(')
+
+        # delete application zip file
+        os.remove(path_to_app)
+        conn.close()
+        DATASENT = False
+        return
+    DATASENT = True
```
If this is solely used by the main window, it should ideally be moved to the main window class.

dhtdht020 commented on gui/ui_wiiloaddialog.ui:
```diff
@@ -0,0 +1,195 @@
+       </property>
+       <widget class="QWidget" name="tab">
+        <attribute name="title">
+         <string>TCP/IP</string>
```
```suggestion
         <string>Network</string>
```

techydude0713 commented:
Thanks for your feedback, here are new changes.
+ ~In the CLI: "send" is now split to "ipsend" and "geckosend."~ Now simplified to send.
+ Enabled safe mode for Send to Wii
+ Prioritize Send to Wii status messages over icon loading (caused progress bar to be -1)
+ Fixed unsendable Wii applications having the "Send to Wii" toggle on after pressing the Download button.
+ CLI: If no arguments, display the help message.

dhtdht020 commented on gui/WiiLoadDialog.py:
```diff
@@ -0,0 +1,110 @@
+        super().__init__(parent)
+        self.setupUi(self)
+        
+        self.setFont(QFont('Helvetica Neue'))
```
This single setFont function slows down dialog initialization by 3845%. Might be because I'm running this on my secondary drive. Anyways, probably can live without it.
```suggestion
```

techydude0713 replied:
Ironically, there was a delay if setFont wasn't called on my Mac, but I think that was due to a bad install of pyside6. I'll remove it.

dhtdht020 commented on xosc_dl.py:
```diff
@@ -454,34 +458,16 @@ def safe_mode(self, state: bool):
-            logging.warning('Invalid IP Address: ' + ip)
-            QMessageBox.warning(self, 'Invalid IP Address', 'This IP address is invalid.')
+        dialog = WiiLoadDialog(self.current_app, parent=self)
+        status = dialog.exec()
```
"status" here is assigned but never used.

dhtdht020 commented on gui/WiiLoadDialog.py:
```diff
@@ -0,0 +1,110 @@
+import logging
+
+from PySide6 import QtCore
+from PySide6.QtCore import QSize, QStorageInfo, QDir, Qt, QEvent
+from PySide6.QtGui import QIcon, QGuiApplication, QFont
+from PySide6.QtWidgets import QDialog, QDialogButtonBox, QListWidgetItem, QMessageBox, QPlainTextEdit
+
+import gui_helpers
+import wiiload
+from gui import ui_WiiLoadDialog
+from utils import resource_path, file_size
+import serial
+import serial.tools.list_ports
```
Some unused imports here
```suggestion
import logging

from PySide6.QtCore import Qt
from PySide6.QtGui import QIcon, QGuiApplication
from PySide6.QtWidgets import QDialog, QMessageBox

import gui_helpers
import wiiload
from gui import ui_WiiLoadDialog
from utils import resource_path
import serial
import serial.tools.list_ports
```

dhtdht020 commented on gui_helpers.py:
```diff
@@ -1,3 +1,12 @@
+import os
+import logging
 from PySide6.QtCore import QSettings
+from PySide6.QtWidgets import QMessageBox
```
Unused imports
```suggestion
from PySide6.QtCore import QSettings
```

dhtdht020 commented on gui/WiiLoadDialog.py:
```diff
@@ -0,0 +1,110 @@
+            self.PortBox.addItem(x.device)
+            self.PortBoxSerial.append((x.vid,x.pid))
+        
+        if (0x403, 0x6001) in self.PortBoxSerial:
+            self.PortBox.setCurrentIndex(self.PortBoxSerial.index((0x403, 0x6001)))
```
Could you comment the meaning of these values along similar ones in code?

dhtdht020 reviewed (commented):
Thanks for applying the changes. Here are a few more suggestions and then I think this should be good to go.

dhtdht020 commented:
Hey, one more thing- I've noticed that if one selects an invalid device:
1. The program is forever stuck trying to send.
2. If the program is exited, the thread trying to send isn't killed, keeping it running in the background forever

Will appreciate if you can look into better handling that.

techydude0713 commented:
> Hey, one more thing- I've noticed that if one selects an invalid device:
> 
> 1. The program is forever stuck trying to send.
> 2. If the program is exited, the thread trying to send isn't killed, keeping it running in the background forever
> 
> Will appreciate if you can look into better handling that.

Got it. What does the status message say when it gets hung up?

dhtdht020 commented:
> > Hey, one more thing- I've noticed that if one selects an invalid device:
> > 
> > 1. The program is forever stuck trying to send.
> > 2. If the program is exited, the thread trying to send isn't killed, keeping it running in the background forever
> > 
> > Will appreciate if you can look into better handling that.
> 
> Got it. What does the status message say when it gets hung up?

"Sending app..."

techydude0713 commented:
I can't reproduce the error, but this _should_ fix it.
Also:
+ I added a timeout when the device opens using new import: func-timeout. (I need to check on Windows, but works good on macOS.)
+ If the app icons are loading while a download is happening or being uploaded, don't send any status messages from download_app_icons.
+ If OSCDL is doing something, block the user from exiting the application, until everything is complete.
+ Other minor GUI improvements

dhtdht020 commented:
Unfortunately did not solve the problem, does not reach time out in this scenario. Note that by "invalid device" I mean a serial device that is alive and connected, but not a USB Gecko.
I am not very familiar with serial communication but could it be possible to determine if the selected device is indeed a USB Gecko? If so, that would solve the issue.

techydude0713 commented:
I limited the search to only find compatible USBGecko Devices. If there are none connected, the combobox is disabled, until the refresh button is clicked.

techydude0713 commented:
I found a method to include extra directories when sending the zip file, including those at root. (Targets Issue #37)
I tested it with WiiSXRX with good results, but I haven't tested others yet.
Should it be added to this PR, or wait until a later time?

dhtdht020 commented:
This is already a pretty big pull request, perhaps at a later time.

dhtdht020 commented:
Is func-timeout still needed?

techydude0713 commented:
> Is func-timeout still needed?

I would recommend keeping it, as this ensures that even if the device hangs, the program won't hang with it.
Without func-timeout on macOS, OSCDL will hang for at least 5 seconds before reporting "Resource busy" with the serial device.

dhtdht020 commented:
> > Is func-timeout still needed?
> 
> I would recommend keeping it, as this ensures that even if the device hangs, the program won't hang with it. Without func-timeout on macOS, OSCDL will hang for at least 5 seconds before reporting "Resource busy" with the serial device.

Is func-timeout the best way to do this? Could the status of the resource be checked ahead of time, to make sure it's not busy?

techydude0713 commented:
> > > Is func-timeout still needed?
> > 
> > 
> > I would recommend keeping it, as this ensures that even if the device hangs, the program won't hang with it. Without func-timeout on macOS, OSCDL will hang for at least 5 seconds before reporting "Resource busy" with the serial device.
> 
> Is func-timeout the best way to do this? Could the status of the resource be checked ahead of time, to make sure it's not busy?

IIRC, the only good way to check is to try and open the serial device. 
While it is possible to [recreate func-timeout from scratch](https://stackoverflow.com/questions/34562473/most-pythonic-way-to-kill-a-thread-after-some-period-of-time) (this is what func-timeout does), it may be better to use func-timeout as it is less code to write, plus it includes an exception for when the function times out.

dhtdht020 commented:
> 



> > > > Is func-timeout still needed?
> > > 
> > > 
> > > I would recommend keeping it, as this ensures that even if the device hangs, the program won't hang with it. Without func-timeout on macOS, OSCDL will hang for at least 5 seconds before reporting "Resource busy" with the serial device.
> > 
> > 
> > Is func-timeout the best way to do this? Could the status of the resource be checked ahead of time, to make sure it's not busy?
> 
> IIRC, the only good way to check is to try and open the serial device. While it is possible to [recreate func-timeout from scratch](https://stackoverflow.com/questions/34562473/most-pythonic-way-to-kill-a-thread-after-some-period-of-time) (this is what func-timeout does), it may be better to use func-timeout as it is less code to write, plus it includes an exception for when the function times out.

Alright, understood.

dhtdht020 commented:
Alright, this looks good.
